OPHD/StructureManager.h

```diff
@@ -112,20 +128,10 @@ class StructureManager
 		auto& table = mComponents[ComponentTy::componentTypeID];
 		auto it = table.find(s);
 		if (it != table.end())
-			return reinterpret_cast<ComponentTy*>(it->second);
+			return static_cast<ComponentTy*>(it->second);
```

Comment:
I would expect either `reinterpret_cast` or `dynamic_cast` here. If there's a chance it could fail, `dynamic_cast` probably makes more sense.

Though being able to avoid casts completely would be better. I'm wondering if that might be possible somehow.

Reply:
reinterpret_cast would be wrong - you'd get the wrong address with multiple inheritance, for instance. dynamic_cast involves RTTI, which is not necessary in this case we know the type. The only reason to use dynamic_cast would be to dynamically discover whether an instance is in fact the type you're querying.

If you really want to avoid casts completely (why?) then I mentioned a possible solution [here](https://github.com/OutpostUniverse/OPHD/pull/836#discussion_r603556121). I do not suggest actually doing that though.

Reply:
I would agree that static_cast is the correct cast to use here. Reinterpret is ... scary. Dynamic is unnecessary if we know the type we're expecting. If it's wring it explodes anyway and the compiler should be able to determine that.

As for avoiding casting... that would be ideal but would require an entire rewrite of much of the code base. I'm not even sure how I'd do it differently other than using a Data Oriented Programming approach over OOP (doable... but tremendous effort would go into it and for the gain of avoiding casts). At least with the static_cast operator we know exactly where we're subverting the type system.

Reply:
That's a good point about `reinterpret_cast` being potentially wrong for multiple inheritance. Though I also tend to avoid multiple inheritance. Or at least, the only times I use multiple inheritance are when working with interfaces, where the problems of multiple inheritance generally go away.

It's pretty rare that I find the need for `dynamic_cast`. Plus `dynamic_cast` may not work if you disable RTTI in compiler settings.

For `static_cast`, I normally think of that when you are actively doing something that requires the compiler to insert code. Converting between `int` and `float` would use a `static_cast`. Or invoking a type conversion member function. With `reinterpret_cast`, there is no compiler inserted code. It just treats the memory as if it was a different compile-time known type. As that seems to be the case here, I would expect `reinterpret_cast` to be appropriate.

Question: What happens if you use `static_cast` on a type that has an overloaded type conversion method? I suspect it would produce the wrong results here, and what you really want is `reinterpret_cast`.

----

I like to avoid casts when possible because they are a way of defeating the type system and removing safety checks and warnings. If you don't really need to do it, then it's best to avoid casts. That's especially true when refactoring, as changes to types can cause casts to do invalid things, and to silence warnings while doing it. Besides, with modern template facilities, you can often get correct type safe behaviour without disabling any warnings.

In this particular case, I'm not entirely sure how to avoid casts. Maybe with a bit of thinking, something could be done. Like say, if there are a fixed number of component types, you could have separate `std::vector` to hold each type.

Reply:
>For static_cast, I normally think of that when you are actively doing something that requires the compiler to insert code.

It won't insert code if there was no need to do so. Static_cast simply ensures you're converting between related types in a correct way. The cases where it will insert code are the cases where it _needs_ to insert code for the cast to be correct.

>As that seems to be the case here, I would expect reinterpret_cast to be appropriate.

attachComponent takes a StructureComponent pointer. If you pass it a pointer to a derived class (which is the intended pattern), then that's implicitly a static cast. Using reinterpret_cast to get back the original pointer is simply wrong. Forbidding use of multiple inheritance makes the invalid code happen to work, but that's a bad reason not to use the correct casting method.

>Question: What happens if you use static_cast on a type that has an overloaded type conversion method? I suspect it would produce the wrong results here, and what you really want is reinterpret_cast.

If the overloaded type conversion method does the wrong thing, then that's bad code.

Reply:
I think you're correct about using [`static_cast`](https://en.cppreference.com/w/cpp/language/static_cast).

According to cppreference, this would be case (2). Now that I think about it, I rarely use `static_cast` outside of case (1). Whenever I see `static_cast`, I pretty much always think of case (1).

When dealing with derived classes, normally being able to call virtual functions means I don't need to worry about down casting. The only conversion I need to worry about is up casting, and that happens implicitly. The few cases where I have restored to down casting have generally been of the kind that requires `dynamic_cast`.

Then there's my habit of resorting to `reinterpret_cast` because of all the patching work I used to do. It has probably become my hammer.

Another interesting source of information was:
https://www.mooclab.club/threads/how-do-you-explain-the-differences-among-static_cast-reinterpret_cast-const_cast-and.4079/

----

As for what would happen if someone overloaded a type conversion method. Probably nothing special. The cast is on the pointer, not on the type.

It occurs to me that my usual use of `static_cast` with case (1) means I'm normally working on the class type, not on a pointer to the class type.